src/kvdb/prometheus.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright 2022 Parity Technologies (UK) Ltd.
+// This file is part of polkadot-introspector.
+//
+// polkadot-introspector is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// polkadot-introspector is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with polkadot-introspector.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::kvdb::IntrospectorKvdb;
+use clap::Parser;
+use color_eyre::Result;
+use log::info;
+use prometheus_endpoint::{prometheus::IntGaugeVec, Opts, Registry};
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+#[derive(Clone, Debug, Parser, Default)]
+#[clap(rename_all = "kebab-case")]
+pub struct KvdbPrometheusOptions {
+	/// Prometheus endpoint port.
+	#[clap(long, default_value = "65432")]
+	port: u16,
+	/// Database poll timeout (default, once per 5 minutes).
+	#[clap(long, default_value = "300.0")]
+	poll_timeout: f32,
+}
+
+struct KvdbPrometheusMetrics {
+	keys_size_gauge: IntGaugeVec,
+	values_size_gauge: IntGaugeVec,
+	elements_count_gauge: IntGaugeVec,
+}
+
+pub async fn run_prometheus_endpoint_with_db<D: IntrospectorKvdb + Send + Sync + 'static>(
+	db: D,
+	prometheus_opts: KvdbPrometheusOptions,
+) -> Result<Vec<tokio::task::JoinHandle<()>>> {
+	let prometheus_registry = Registry::new_custom(Some("introspector".into()), None)?;
+	let metrics = Arc::new(Mutex::new(register_metrics(&prometheus_registry)));
+	let socket_addr =
+		std::net::SocketAddr::new(std::net::IpAddr::V4(std::net::Ipv4Addr::new(0, 0, 0, 0)), prometheus_opts.port);
+	let mut futures: Vec<tokio::task::JoinHandle<()>> = vec![];
+	futures.push(tokio::spawn(async move {
+		prometheus_endpoint::init_prometheus(socket_addr, prometheus_registry)
+			.await
+			.unwrap()
+	}));
+	futures.push(tokio::spawn(async move {
+		update_db(db, metrics.clone(), prometheus_opts).await;
+	}));
+	info!("Starting prometheus node on {:?}", socket_addr);
+
+	Ok(futures)
+}
+
+struct UpdateResult {
+	column: String,
+	keys_count: i64,
+	keys_space: i64,
+	values_space: i64,
+}
+
+async fn update_db<D: IntrospectorKvdb>(
+	db: D,
+	metrics: Arc<Mutex<KvdbPrometheusMetrics>>,
+	prometheus_opts: KvdbPrometheusOptions,
+) {
+	loop {
+		info!("Starting update db iteration");
+		let columns = db.list_columns().unwrap();
+		let mut update_results: Vec<UpdateResult> = vec![];
+		for col in columns {
+			let mut keys_space = 0_i64;
+			let mut keys_count = 0_i64;
+			let mut values_space = 0_i64;
+
+			info!("Iterating over column {}", col.as_str());
+			let iter = db.iter_values(col.as_str()).unwrap();
```

Comment:
Same here

Reply:
That's a good point, as db can be indeed temporarily unavailable. However, as this db is not reopened, there is no point in not exiting here I think.